### PR TITLE
fix(mongo-upgrader): make idempotent and consider prefix - 4.4.x

### PR DESCRIPTION
**Pull request by phiz71**
## Issue

https://gravitee.atlassian.net/browse/APIM-5530

## Description

Mongo upgraders have to use the prefix of tables if it exists.
In order to restart the upgraders, we need to set a version.
An update has been made on the upgrader fwk to allow upgrader versioning.

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-kziianazzq.chromatic.com)

### CODE REVIEW

**Review by guillaumelamirand — APPROVED** [review]

**Review by ytvnr — APPROVED** [review]

File: gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/environment/MissingEnvironmentUpgrader.java [hunk 1]
```diff
@@ -45,45 +51,43 @@ public boolean upgrade() {
...
-            .getCollection("subscriptions")
+        this.getCollection("subscriptions")
             .find()
             .projection(projection)
             .forEach(subscription ->
                 bulkActions.add(
                     new UpdateManyModel<>(
-                        Filters.in("subscriptions", subscription.getString("_id")),
-                        new Document("$set", new Document("environmentId", subscription.getString("environmentId")))
+                        Filters.and(Filters.eq("subscriptions", subscription.getString("_id")), Filters.exists("environmentId", false)),
```
Comment (guillaumelamirand) [inline gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/environment/MissingEnvironmentUpgrader.java]: Filters.in 
```suggestion
                        Filters.and(Filters.in("subscriptions", subscription.getString("_id")), Filters.exists("environmentId", false)),
```

File: gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/environment/MissingEnvironmentUpgrader.java [hunk 2]
```diff
@@ -45,45 +51,43 @@ public boolean upgrade() {
...
+                        Updates.set("environmentId", subscription.getString("environmentId"))
```
Comment (guillaumelamirand) [inline gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/environment/MissingEnvironmentUpgrader.java:84]: Don't you miss the exist filter?

File: gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/plan/PlanDefinitionVersionUpgrader.java [hunk 3]
```diff
@@ -32,27 +34,31 @@ public class PlanDefinitionVersionUpgrader extends MongoUpgrader {
...
+        this.getCollection("apis")
             .find(query)
             .projection(projection)
             .forEach(v4Api -> {
                 bulkActions.add(
                     new UpdateManyModel<>(
-                        new Document("api", v4Api.getString("_id")),
-                        new Document("$set", new Document("definitionVersion", DefinitionVersion.V4.name()))
+                        Filters.and(Filters.eq("api", v4Api.getString("_id")), Filters.exists("definitionVersion", false)),
+                        Updates.set("definitionVersion", DefinitionVersion.V4.name())
```
Comment (guillaumelamirand) [inline gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/plan/PlanDefinitionVersionUpgrader.java:55]: Don't you miss the exist filter?

**Review by guillaumelamirand — CHANGES_REQUESTED** [review]
Few comments

File: gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/environment/MissingEnvironmentUpgrader.java (hunk 2 above)
Comment (phiz71) [reply gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/environment/MissingEnvironmentUpgrader.java:84]: it is here
`, Filters.exists("environmentId", false)`

File: gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/plan/PlanDefinitionVersionUpgrader.java (hunk 3 above)
Comment (phiz71) [reply gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/plan/PlanDefinitionVersionUpgrader.java:55]: it is here
`, Filters.exists("environmentId", false)`

**Review by guillaumelamirand — APPROVED** [review]

**Pull request merged**